lib/chef/cookbook/cookbook_version_loader.rb

```diff
@@ -212,6 +224,24 @@ def chefignore
         @chefignore ||= Chefignore.new(File.basename(cookbook_path))
       end
 
+      def load_all_unignored_files
+        # List all directories that are not dotdirs, then list everything in
+        # those, *including* dotfiles and dotdirs.
+        # Finally, list all the _files_ at the root, including dotfiles, and
+        # merge those in.
+        Dir.glob(File.join(Chef::Util::PathHelper.escape_glob(cookbook_path), "*")).each do |dir|
+          dir = Chef::Util::PathHelper.cleanpath(dir)
+          next unless File.directory?(dir)
+
+          Dir.glob(File.join(Chef::Util::PathHelper.escape_glob(dir), "**/*"), File::FNM_DOTMATCH).each do |file|
+            name = Chef::Util::PathHelper.relative_path_from(@cookbook_path, file)
+            cookbook_settings[:all_unignored_files][name] = file
+          end
+        end
+        load_root_files
+        cookbook_settings[:all_unignored_files].merge!(cookbook_settings[:root_filenames])
+      end
```

Comment:
I think I'd start more like this:

```ruby
def load_all_files
  root_entries = Dir.glob(File.join(Chef::Util::PathHelper.escape_glob_dir(cookbook_path), "*"), File::FNM_DOTMATCH).reject { |f|  File.basename(f).start_with?(".") && File.directory?(f) }

  root_entries.each do |path|
    Dir.glob(File.join(Chef::Util::PathHelper.escape_glob_dir(path), "**/*"), File::FNM_DOTMATCH).each do |file|
      name = Chef::Util::PathHelper.relative_path_from(@cookbook_path, file)
      cookbook_settings[:all_files][name] = file
    end
  end
end
```

Then instead of loading the root_files in you could go the other way and extract the root_files from the all_files if you want to optimize...